### PR TITLE
Setup container for docker provisioning

### PR DESCRIPTION
**Pull request by taimir**
### WIP: Work In Progress, do not merge

@zreigz @batikanu This is the first draft for the provisioning mechanism proposed by @cheld. I have so far tested the `setup-master.sh` on my local machine and it does what it is expected to do. I'm posting this PR so that you guys can already express any ideas / concerns and to avoid unnecessary overlaps if someone already has started work on this.

I'll be back on Wednesday. @zreigz if you want to work on this, feel free to modify this branch :) that's the reason why I pushed it to the common repository.

TODO: add all the `.sh` scripts directly to the hyperkube container and do not mount them
TODO: replace `taimir93/hyperkube` with the official modified hyperkube
TODO: test this in a multi-node setup and ensure that the containers can connect to each other over the flannel overlay
TODO: add the option for a private docker registry
TODO: wire the `setup_master.sh` and `setup_minion.sh` scripts to the heat templates

EDIT: `debug/clean.sh` resets the state of the machine you're deploying on. All the files in the `debug` folder are just for convenience (dev help), they have nothing to do with the actual provisioning.

### CODE REVIEW

File: setup-container/setup/setup_master.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    --net=host \
+    --pid=host \
+    --privileged \
+    taimir93/hyperkube:1 \
+    /setup/install_cni.sh
+
+echo "Installing kubernetes ..."
+docker run -d \
+    --volume=/:/hostfs:rw \
+    --volume=`pwd`:/setup:ro \
```
Comment (cheld) [inline setup-container/setup/setup_master.sh]: this is a hack. the script will be added to hyperkube. Atanas will do on wednesday

File: setup-container/setup/setup_master.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    -v /dev/net:/dev/net \
+    quay.io/coreos/flannel:${FLANNEL_VERSION} \
+    /opt/bin/flanneld \
+        --ip-masq="${FLANNEL_IPMASQ}" \
+        --iface="${FLANNEL_IFACE}"
+
+echo "Installing CNI..."
+docker run -d \
+    --volume=/:/hostfs:rw \
+    --volume=`pwd`:/setup:ro \
```
Comment (cheld) [inline setup-container/setup/setup_master.sh]: temporary hack. Will be refactored

File: setup-container/hyperkube/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#
+# Creates hyperkube with CNI binaries and flannel configuration
```
Comment (taimir) [inline setup-container/hyperkube/Dockerfile]: @zreigz I copied this from you pull request, temporarily until your changes are merged

**Comment by taimir** [comment]
So just to keep everyone updated: I just tried the `setup-master.sh` script on a node in our Openstack installation, and it works fine so far. I can access dashboard, no pods errors, etc.

However when I did try it in a vagrant environment with virtualbox, something wasn't quite right. This should be investigated a bit in the future, and we should also check if the `hyperkube` docker provisioner works there, as well.

I'll proceed with provisioning a minion now.

**Comment by zreigz** [comment]
I was testing hypercube on vagrant and it works. We have project for it: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kube-deploy-vagrant

**Comment by taimir** [comment]
Is that with `flannel + cni` networking or not? I'll have a look at it when I get the minion to run on openstack, maybe I'm missing something fundamental.

**Comment by zreigz** [comment]
Yes this is test environment for PR where I use flannel and cni

File: setup-container/setup/install_kubernetes.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+# copy hyperkube binary
+cp /hyperkube /hostfs/usr/bin/
+
+# copy kubelet service configuration
+cp /setup/kubelet.service /etc/systemd/system/
+
+if [ "$node" == "master" ]; then
+  # clean up old configuration
+  rm -rf /hostfs/var/lib/kubelet
```
Comment (cheld) [inline setup-container/setup/install_kubernetes.sh:16]: I think we do this in the wrapper script. I assume this is the better place for cleanup

**Comment by taimir** [comment]
@zreigz I'm testing these scripts in an OpenStack environment, and I just stumbled upon a connectivity issue. I see that in the heat templates, we have the following security groups:

``` yaml
  secgroup_base:
    type: OS::Neutron::SecurityGroup
    properties:
      rules:
        - protocol: icmp
        - protocol: tcp
          port_range_min: 22
          port_range_max: 22
        - remote_mode: remote_group_id

  secgroup_master:
    type: OS::Neutron::SecurityGroup
    properties:
      rules:
        - protocol: tcp # api-server
          port_range_min: 6443
          port_range_max: 6443

  secgroup_node:
    type: OS::Neutron::SecurityGroup
    properties:
      rules:
        - protocol: tcp
          port_range_min: 30000
          port_range_max: 32767
```

It seems I cannot access the `8080` port on the master node, but that's where the API server is running. However, when I do HTTPS on port `6443`, I can access it.

Where does the port mapping between `8080` and `6443` happen? And if we can't actually access port  `8080`, why do I see in our docker provisioner that the minion's kubelet is started with API-Server on 8080:

``` sh
docker run \
        --name=kubelet \
        --volume=/:/rootfs:ro \
        --volume=/sys:/sys:ro \
        --volume=/var/lib/docker/:/var/lib/docker:rw \
        --volume=/var/run:/var/run:rw \
        --volume=/var/lib/kubelet:/var/lib/kubelet:shared \
        --net=host \
        --pid=host \
        --privileged=true \
        --restart=on-failure \
        -d \
        ${HYPERKUBE_IMAGE} \
        /hyperkube kubelet \
            --hostname-override=${NODE_IP} \
            --address="0.0.0.0" \
            --api-servers=http://${MASTER_IP}:8080 \
            --cluster-dns=10.0.0.10 \
            --cluster-domain=cluster.local \
            --allow-privileged=true --v=2  \
            --pod-infra-container-image=${PAUSE_IMAGE}
```

**Comment by taimir** [comment]
Is this some discrepancy between the heat scripts (which configure a secure networking environment) and the way the api-server is started in the docker-provisioner currently? Or there's something more involved in the networking setup?

**Comment by zreigz** [comment]
I think there are some default Access & Security rules in OpenStack which allow access any port. Probably it depends on Openstack installation. In devstack installation everything is opened and allowed

**Comment by taimir** [comment]
Hey, just for documentation purposes: the connectivity issue was not at all related to OpenStack, I manually checked all the security groups and `iptables` rules. It has to do with the behavior of `hyperkube`, more specifically with the manifests it contains. The API server (in `/manifests/master.json`) is started with:

``` json
"name": "apiserver",
      "image": "gcr.io/google_containers/hyperkube-amd64:v1.3.0-beta.1",
      "command": [
              "/hyperkube",
              "apiserver",
              "--service-cluster-ip-range=10.0.0.1/24",
              "--insecure-bind-address=127.0.0.1",
              "--etcd-servers=http://127.0.0.1:4001",
              "--admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,ResourceQuota",
              "--client-ca-file=/srv/kubernetes/ca.crt",
              "--basic-auth-file=/srv/kubernetes/basic_auth.csv",
              "--min-request-timeout=300",
              "--tls-cert-file=/srv/kubernetes/server.cert",
              "--tls-private-key-file=/srv/kubernetes/server.key",
              "--token-auth-file=/srv/kubernetes/known_tokens.csv",
              "--allow-privileged=true",
              "--v=2"
      ]
```

`localhost:8080` can thus only be used locally, and `${MASTER_IP}:6443` is the official IP of the server (the secure connection defaults to port 6443, if I'm not mistaken) for connections from other nodes. `--insecure-bind-address` basically says that only the master node can access the API server insecurely.

**Comment by cheld** [comment]
@taimir this is correct. In my understanding it is not so easy to enable tls between master and minions without salt. So we should start without and think about it later

**Comment by cheld** [comment]
BTW: if I remember correctly I have seen a proposal for bootstrapping root certificates in the cluster

**Comment by taimir** [comment]
OK, so after my first multi-node experiment, here's the results in summary:
- With the setup scripts above, we can successfully create a functional multi-node kubernetes cluster. The only dependency is `centos + docker`, and `systemd` if we start the kubelet as a service.
- For the connectivity to work out, I had to use the `/etc/kubernetes/manifests-multi` from hyperkube. Unfortunately, they do not solve the certificates exchange between the master and minion, just enable insecure access to the API server from any node.
- I've tested this on Packstack liberty, `centos+docker` VMs. Flannel networking + cni works fine.

DONE: check this in a vagrant multi-node setup
TODO: add private registry option to the scripts (like in the original docker provisioner)
TODO: restrict the insecure access to master to nodes that are only in the cluster with `iptables`. Later find a way to deal with the certificates in a clean way (in another PR)

File: setup-container/setup/kubelet.service [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+Description=Kubernetes Kubelet
+Documentation=https://github.com/kubernetes/kubernetes
+
+[Service]
+ExecStartPre=/usr/bin/mkdir -p /etc/kubernetes/manifests
+ExecStart=/usr/bin/hyperkube kubelet \
+    --allow-privileged=true \
+    --hostname-override=127.0.0.1 \
+    --address=0.0.0.0 \
+    --api-servers=http://localhost:8080 \
```
Comment (taimir) [inline setup-container/setup/kubelet.service:11]: TODO: figure out how to pass the Master IP in the service definition when we provision a minion.

Comment (zreigz) [inline setup-container/setup/kubelet.service:11]: On master/minion host you have /etc/sysconfig/heat-params file with environment variables like:

```
      MASTER_IP
      ROLE
      DOCKER_REGISTRY_URL
      DOCKER_REGISTRY_PREFIX
```

Comment (cheld) [inline setup-container/setup/kubelet.service:11]: Heat should pass the parameters to the script. The script should not have a dependency to heat

Comment (cheld) [inline setup-container/setup/kubelet.service:11]: I assume the proper way is to source a config file e.g. /etc/Kubernetes..... from unit file. Check Docker.servuce or salt Scripts

**Comment by taimir** [comment]
Closing this WIP, the official PR is https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/k8s-docker-provisioner/pull/35

**Pull request closed without merging**